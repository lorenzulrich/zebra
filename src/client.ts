'use client';

export { default as BackendIncludes } from './server/components/client/BackendIncludes';
