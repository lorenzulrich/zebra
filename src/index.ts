export * from './types';

export { default as NeosRevalidate } from './api/revalidate';

export { default as NeosContext } from './utils/context';

export {
  injectNeosBackendMetadata,
  nodeMetadata,
  loadStaticPaths,
  loadStaticProps,
  loadServerSideDocumentProps,
  loadServerSideNodeProps,
  convertApiUrlToPath,
  withZebra
} from './utils/helper';

export { useNode, useDocumentNode, useSiteNode, useMeta, useInBackend } from './utils/hooks';

export { default as Preview } from './lib/pages/Preview';
export { default as Frontend } from './lib/pages/Frontend';

export { default as ChildNodes } from './lib/components/ChildNodes';
export { default as ContentCollection } from './lib/components/ContentCollection';
export { default as ContentComponent } from './lib/components/ContentComponent';
export { default as ContentRegistry } from './lib/components/ContentRegistry';
export { default as Editable } from './lib/components/Editable';
export { default as BackendContainer } from './lib/components/BackendContainer';
