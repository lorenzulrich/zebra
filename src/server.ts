export { default as ContentCollection } from './server/components/ContentCollection';
export { default as ContentCollectionProvider } from './server/components/ContentCollectionProvider';
export { default as ContentComponent } from './server/components/ContentComponent';
export { default as ContentComponentProvider } from './server/components/ContentComponentProvider';
export { default as Editable } from './server/components/Editable';
export { default as NodeRenderer } from './server/components/NodeRenderer';
export { NeosServerContext } from './server/utils/context';
export {
  loadDocumentProps,
  loadDocumentPropsCached,
  loadPreviewDocumentProps,
  loadPreviewDocumentPropsCached,
  loadSiteProps,
} from './server/utils/dataLoader';
export { resolveCurrentNode, resolveCurrentNodeRecursive } from './server/utils/helper';
export {
  useContentCollection,
  useContentComponent,
  useDocumentNode,
  useEditPreviewMode,
  useInBackend,
  useMeta,
  useNode,
  useSiteNode,
} from './server/utils/hooks';
export { getNodeType, initNodeTypes } from './server/utils/nodeTypes';
