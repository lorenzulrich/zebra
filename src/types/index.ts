export interface NeosData {
  node: NeosContentNode;
  site: NeosNode;
  meta: any;
  backend?: BackendProps;
}

export interface NeosNode {
  identifier: string;
  nodeType: string;
  contextPath: string;
  properties: {
    [key: string]: any;
  };
  label: string;
  depth: number;
  removed: boolean;
  creationDateTime: string;
  lastModificationDateTime: string;
  lastPublicationDateTime: string;
  path: string;
  hidden: boolean;
  hiddenBeforeDateTime: string;
  hiddenAfterDateTime: string;
  parent: string;
  isAutoCreated: boolean;
  matchesCurrentDimensions: boolean;
  isFullyLoaded: boolean;
}

export interface NeosDocumentNode extends NeosNode {
  children?: NeosContentNode[];
}

export interface NeosContentNode extends NeosNode {
  nodeName: string;
  children?: NeosContentNode[];
  backend?: {
    serializedNode: any;
  };
}

export interface BackendInclude {
  key: string;
  type: 'script' | 'link';
  content?: string;
  src?: string;
  href?: string;
  rel?: string;
}

export interface BackendProps {
  documentInformation?: any;
  editPreviewMode?: BackendEditPreviewMode;
  guestFrameApplication?: BackendInclude[];
}

export interface BackendEditPreviewMode {
  name: string;
  isPreview: boolean;
  isEdit: boolean;
  options: Record<string, any> | null;
}

export type NeosNodeTypes = Record<string, React.FC>;

export interface NeosContextProps {
  node: NeosContentNode;
  documentNode: NeosDocumentNode;
  site: NeosNode;
  meta: any;
  nodeTypes: Record<string, any>;
  inBackend: boolean;
  editPreviewMode?: BackendEditPreviewMode;
}

export interface DocumentsResponse {
  documents: DocumentsItem[];
}

export interface DocumentsItem {
  identifier: string;
  contextPath: string;
  dimensions: Record<string, string[]>;
  site: string;
  meta: {
    title: string;
  };
  routePath: string;
  renderUrl: string;
  creationDateTime: string;
  lastPublicationDateTime?: string;
}

export interface ApiErrors {
  errors?: { message: string; code: number }[];
}
