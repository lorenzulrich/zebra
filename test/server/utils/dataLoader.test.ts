import { headers } from 'next/headers';
import { afterEach, beforeEach, describe, expect, it, test, vi } from 'vitest';

import {
  loadDocumentProps,
  loadDocumentPropsCached,
  loadPreviewDocumentProps,
  loadSiteProps,
} from '../../../src/server';
import { DataLoaderOptions } from '../../../src/types';

describe('loadDocumentProps', () => {
  it('should throw an error if NEOS_BASE_URL is not set', async () => {
    await expect(loadDocumentProps({ slug: 'foo' })).rejects.toThrowError('Missing NEOS_BASE_URL environment variable');
  });

  describe('with NEOS_BASE_URL set', () => {
    beforeEach(() => {
      vi.stubEnv('NEOS_BASE_URL', 'http://neos:1234');
    });

    describe.each<{ opts?: DataLoaderOptions; expectedFetchConfig: RequestInit }>([
      // No options
      { expectedFetchConfig: { cache: 'no-store', headers: {}, next: undefined } },
      // Specify cache
      { opts: { cache: 'default' }, expectedFetchConfig: { cache: 'default', headers: {}, next: undefined } },
      // Specify cache and next tags
      {
        opts: { cache: 'force-cache', next: { tags: ['neos'] } },
        expectedFetchConfig: { cache: 'force-cache', headers: {}, next: { tags: ['neos'] } },
      },
    ])('with options $opts', ({ opts, expectedFetchConfig }) => {
      it('should fetch from configured API', async () => {
        const fetch = vi.fn().mockResolvedValue(createOkayFetchResponse({ meta: { title: 'Foo' } }));
        vi.stubGlobal('fetch', fetch);

        await expect(loadDocumentProps({ slug: 'foo' }, opts)).resolves.toStrictEqual({
          meta: { title: 'Foo' },
        });

        expect(fetch).toHaveBeenCalledWith(
          'http://neos:1234/neos/content-api/document?path=%2Ffoo',
          expectedFetchConfig
        );
      });
    });
  });

  describe('with optional option', () => {
    test('should not throw an error if NEOS_BASE_URL is not set', async () => {
      await expect(loadDocumentProps({ slug: 'foo' }, { optional: true })).resolves.toBeUndefined();
    });
  });
});

describe('loadDocumentPropsCached', () => {
  beforeEach(() => {
    vi.stubEnv('NEOS_BASE_URL', 'http://neos:1234');
  });

  test('should cache the result', async () => {
    const fetch = vi.fn().mockResolvedValue(createOkayFetchResponse({ meta: { title: 'Foo' } }));
    vi.stubGlobal('fetch', fetch);

    await expect(loadDocumentPropsCached('foo')).resolves.toStrictEqual({
      meta: { title: 'Foo' },
    });

    await expect(loadDocumentPropsCached('foo')).resolves.toStrictEqual({
      meta: { title: 'Foo' },
    });

    expect(fetch).toHaveBeenCalledTimes(1);
  });

  test('should cache the result again', async () => {
    const fetch = vi.fn().mockResolvedValue(createOkayFetchResponse({ meta: { title: 'Foo' } }));
    vi.stubGlobal('fetch', fetch);

    await expect(loadDocumentPropsCached('foo')).resolves.toStrictEqual({
      meta: { title: 'Foo' },
    });

    await expect(loadDocumentPropsCached('foo')).resolves.toStrictEqual({
      meta: { title: 'Foo' },
    });

    expect(fetch).toHaveBeenCalledTimes(1);
  });
});

describe('loadPreviewDocumentProps', () => {
  describe('with NEOS_BASE_URL set', () => {
    beforeEach(() => {
      vi.stubEnv('NEOS_BASE_URL', 'http://neos:1234');
    });

    // TODO Add test without context path query param
    // TODO Add test for PUBLIC_BASE_URL env and proxy headers

    describe('with Neos session cookie', () => {
      beforeEach(() => {
        vi.mocked(headers).mockReturnValue(new Headers({ Cookie: 'a-session-cookie' }));
      });

      describe.each<{ opts?: DataLoaderOptions }>([
        // No options
        {},
        // Override cache, should have no effect for preview
        { opts: { cache: 'default' } },
      ])('with options $opts', ({ opts }) => {
        it('should fetch from configured API', async () => {
          const fetch = vi.fn().mockResolvedValue(createOkayFetchResponse({ meta: { title: 'Foo Preview' } }));
          vi.stubGlobal('fetch', fetch);

          await expect(
            loadPreviewDocumentProps({ 'node[__contextNodePath]': 'foo/bar@user-me' }, opts)
          ).resolves.toStrictEqual({
            meta: { title: 'Foo Preview' },
          });

          expect(fetch).toHaveBeenCalledWith(
            'http://neos:1234/neos/content-api/document?contextPath=foo%2Fbar%40user-me',
            {
              cache: 'no-store',
              headers: {
                Cookie: 'a-session-cookie',
              },
              next: undefined,
            }
          );
        });
      });
    });
  });
});

describe('loadSiteProps', () => {
  it('should throw an error if NEOS_BASE_URL is not set', async () => {
    await expect(loadSiteProps()).rejects.toThrowError('Missing NEOS_BASE_URL environment variable');
  });

  describe('with NEOS_BASE_URL set', () => {
    beforeEach(() => {
      vi.stubEnv('NEOS_BASE_URL', 'http://neos:1234');
    });

    describe.each<{ opts?: DataLoaderOptions; expectedFetchConfig: RequestInit }>([
      // No options
      { expectedFetchConfig: { cache: 'no-store', headers: {}, next: undefined } },
      // Override cache
      { opts: { cache: 'default' }, expectedFetchConfig: { cache: 'default', headers: {}, next: undefined } },
    ])('with options $opts', ({ opts, expectedFetchConfig }) => {
      it('should fetch from configured API', async () => {
        const fetch = vi.fn().mockResolvedValue(createOkayFetchResponse({ meta: { title: 'Foo' } }));
        vi.stubGlobal('fetch', fetch);

        await expect(loadSiteProps(opts)).resolves.toStrictEqual({
          meta: { title: 'Foo' },
        });

        expect(fetch).toHaveBeenCalledWith('http://neos:1234/neos/content-api/site', expectedFetchConfig);
      });
    });
  });

  describe('with optional option', () => {
    test('should not throw an error if NEOS_BASE_URL is not set', async () => {
      await expect(loadSiteProps({ optional: true })).resolves.toBeUndefined();
    });
  });
});

// --- Mock implementations ---

let reactCached = false;

// Mock the `cache` export from React:
// 1. it is not exported in the _normal_ react version (using the canary channel would work though), but inside Next.js we can use it.
// 2. we don't want to use the real cache, because we can not control it in tests.
vi.mock('react', async (importOriginal) => {
  const mod = await importOriginal<typeof import('react')>();
  return {
    ...mod,
    // The cache function is a bit tricky, since it is used at the module level when importing the module,
    // so we have to specify the implementation here and now - but still enable a reset of cached values in tests.
    // eslint-disable-next-line @typescript-eslint/ban-types
    cache: vi.fn(<CachedFunction extends Function>(fn: CachedFunction): CachedFunction => {
      // Cache the first call and return the cached value for all subsequent calls.
      let result: any;
      return ((...args: any[]) => {
        if (reactCached) {
          return result;
        }
        reactCached = true;
        return (result = fn(...args));
      }) as any;
    }),
  };
});

vi.mock('next/headers', async (importOriginal) => {
  const mod = await importOriginal<typeof import('next/headers')>();
  return {
    ...mod,
    headers: vi.fn(),
  };
});

afterEach(() => {
  vi.unstubAllEnvs();
  vi.unstubAllGlobals();
  vi.resetAllMocks();
  // Make sure React cache is reset after each test
  reactCached = false;
});

function createOkayFetchResponse(data: any) {
  return { ok: true, status: 200, json: () => new Promise((resolve) => resolve(data)) };
}
