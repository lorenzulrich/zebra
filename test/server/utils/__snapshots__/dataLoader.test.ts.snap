// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`loadDocumentProps > with NEOS_BASE_URL set > with error as JSON > should throw an error with the response status text 1`] = `[Error: Content API responded with errors, status: 500, url: http://neos:1234/neos/content-api/document?path=%2Ffoo, errors: Some error]`;

exports[`loadDocumentProps > with NEOS_BASE_URL set > with error not as JSON > should throw an error with the response status text 1`] = `[Error: Content API responded with unexpected error, status: 500, url: http://neos:1234/neos/content-api/document?path=%2Ffoo]`;

exports[`loadDocumentProps > with NEOS_BASE_URL set > with ok but not as JSON > should throw an error with the response status text 1`] = `[Error: Content API responded with invalid JSON: SyntaxError: Unexpected token '<', "<html><bod"... is not valid JSON, status: 200, url: http://neos:1234/neos/content-api/document?path=%2Ffoo]`;

exports[`loadPreviewDocumentProps > with NEOS_BASE_URL set > with Neos session cookie > with error as JSON > should throw an error with the response status text 1`] = `[Error: Content API responded with errors, status: 500, url: http://neos:1234/neos/content-api/document?contextPath=foo%2Fbar%40user-me, errors: Some error]`;

exports[`loadPreviewDocumentProps > with NEOS_BASE_URL set > with Neos session cookie > with error not as JSON > should throw an error with the response status text 1`] = `[Error: Content API responded with unexpected error, status: 500, url: http://neos:1234/neos/content-api/document?contextPath=foo%2Fbar%40user-me]`;

exports[`loadSiteProps > with NEOS_BASE_URL set > with error as JSON > should throw an error with the response status text 1`] = `[Error: Content API responded with errors, status: 500, url: http://neos:1234/neos/content-api/site, errors: Some error]`;

exports[`loadSiteProps > with NEOS_BASE_URL set > with error not as JSON > should throw an error with the response status text 1`] = `[Error: Content API responded with unexpected error, status: 500, url: http://neos:1234/neos/content-api/site]`;
